### ISSUE TITLE
RPC Client doesn't gracefully close Websocket Connections

### ISSUE THREAD

**Issue by moorow** [ethereum/go-ethereum#30482]
#### System information

Geth version: `1.14.8`
OS & Version: OSX

#### Expected behaviour

When rpc.Client Close() is called, I would expect a message like this to be sent

```
err := cli.writeMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
```

#### Actual behaviour

The client closes the TCP conn without sending a WS Close message, leading to `websocket: close 1006 (abnormal closure): unexpected EOF` on the server side.

#### Steps to reproduce the behaviour

Open a WS connection with the rpc.Client. Close the connection.

**Comment by holiman** [comment]
Odd. In `rpc/client.go`, the `dispatch` loop calls `close` on the underlying conn

```golang
	defer func() {
		close(c.closing)
		if reading {
			conn.close(ErrClientQuit, nil)
			c.drainRead()
		}
		close(c.didClose)
	}()
```
At some point, I'd have expected that to trickle down to the  `gorilla/websocket` conn, which has the following default close-handler: 
```golang
func (c *Conn) SetCloseHandler(h func(code int, text string) error) {
	if h == nil {
		h = func(code int, text string) error {
			message := FormatCloseMessage(code, "")
			c.WriteControl(CloseMessage, message, time.Now().Add(writeWait))
			return nil
		}
	}
	c.handleClose = h
}
```

**Comment by moorow** [comment]
So `conn.close(ErrClientQuit, nil)` eventually calls close on the jsonCodec

```
func (c *jsonCodec) close() {
	c.closer.Do(func() {
		close(c.closeCh)
		c.conn.Close()
	})
}
``` 

This in turn calls Close on the Gorrillas WS conn as you say.

```
// Close closes the underlying network connection without sending or waiting
// for a close message.
func (c *Conn) Close() error {
	return c.conn.Close()
}
```

This function unfortunately just closes the TPC connection.

The Close Handler only gets called if a CloseMessage is received.

**Comment by holiman** [comment]
Hm, it seems this is a long-standing issue in gorilla/websocket: https://github.com/gorilla/websocket/issues/448

**Comment by moorow** [comment]
Yeah not sure if they are going to include or if they expect devs to implement themselves. When I use gorilla/websocket I tend to do something like this to solve the issue:

```
func (cli *Client) close() {
	// Gracefully close the connection by sending a close message
	if err := cli.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")); err != nil {
              //handle err
	}

	// Wait for the server to acknowledge the close
	select {
	case <-cli.done:
	case <-time.After(2 * time.Second):
	
            //Close the underlying connection
            if err := cli.conn.Close(); err != nil {
	            //handle err
            }
	}
}
```

Unfortunately this isn't possible with rpc Client as I don't have access to the WS conn itself.

**Comment by V0YD23** [comment]
@moorow  is this issue still open ?

**Comment by moorow** [comment]
@V0YD23 as far as I know, I haven't tested recently. I will give it a test tomorrow and update this

**Comment by V0YD23** [comment]
@moorow thanx for the reply and yes if it will not be open  can you suggest me anyother issue on which i can work upon being beginner in open source but have experience in golang and been reading about web3 recently

**Comment by sohailblockchain** [comment]
This may be related to how the websocket subscription lifecycle is handled in rpc/client.go.

If the connection is closed unexpectedly without proper context cancellation, the underlying goroutine may remain open.

Have you tested with explicit client.Close() and ensured the subscription is unsubscribed before closing the connection?

It might also help to enable debug logging using:

--verbosity 4

to see how the RPC layer handles connection teardown.
